java/arcs/sdk/android/storage/ServiceStore.kt

```diff
@@ -103,13 +103,21 @@ class ServiceStore<Data : CrdtData, Op : CrdtOperation, ConsumerData>(
     private val scope = CoroutineScope(coroutineContext)
     private var storageService: IStorageService? = null
     private var serviceConnection: StorageServiceConnection? = null
-    private var channel = Channel<suspend () -> Unit>(Channel.UNLIMITED)
+    private var channel: Channel<suspend () -> Unit>? = null
+    private var flow: Flow<suspend () -> Unit>? = null
 
     init {
         lifecycle.addObserver(this)
-        channel.consumeAsFlow()
-            .onEach { it() }
-            .launchIn(scope)
+        initChannel()
+    }
+
+    // Channel has an internal queue which can retain work if stopped
+    // So we need to create fresh instances when off() invoked
+    private fun initChannel() {
+        channel?.let { it.cancel() }
+        channel = Channel<suspend () -> Unit>(Channel.UNLIMITED)
+        flow = channel!!.consumeAsFlow().onEach { it() }
+        flow!!.launchIn(scope)
```

Comment:
I feel like these operations should happen within a mutex?

Reply:
I think you're right, given that on()/off() can be invoked from multiple threads.